[SIMD] B3/OMG Support

justinmichaud opened the pull request:
#### 8a018265fb90726b8b3e8e0fb88ec9238ee0255c
<pre>
[SIMD] B3/OMG Support
<a href="https://bugs.webkit.org/show_bug.cgi?id=246347">https://bugs.webkit.org/show_bug.cgi?id=246347</a>

Reviewed by Mark Lam.

Support B3/ the OMG tier for SIMD functions.

We add SimdValue and Const128Value as two new B3 value types, plus
opcodes for every wasm simd op.

We also add extra testing modes to simd spec tests to ensure that
the OMG tier is being fully exercised. This should be a temporary
measure until SIMD no longer requires tiering up from LLInt to BBQ.

Finally, we work around a bug in register allocation for SIMD functions
by disabling the use of ARM64 fp callee-saves.

* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::notifyFunctionUsesSIMD):
(JSC::Wasm::B3IRGenerator::addConstant):
(JSC::Wasm::B3IRGenerator::addExtractLane):
(JSC::Wasm::B3IRGenerator::addReplaceLane):
(JSC::Wasm::B3IRGenerator::addSIMDI_V):
(JSC::Wasm::B3IRGenerator::addSIMDV_V):
(JSC::Wasm::B3IRGenerator::addSIMDBitwiseSelect):
(JSC::Wasm::B3IRGenerator::addSIMDRelOp):
(JSC::Wasm::B3IRGenerator::addSIMDSwizzle):
(JSC::Wasm::B3IRGenerator::addSIMDV_VV):
(JSC::Wasm::B3IRGenerator::addSIMDLoad):
(JSC::Wasm::B3IRGenerator::addSIMDStore):
(JSC::Wasm::B3IRGenerator::addSIMDSplat):
(JSC::Wasm::B3IRGenerator::addSIMDShift):
(JSC::Wasm::B3IRGenerator::addSIMDExtmul):
(JSC::Wasm::B3IRGenerator::addSIMDShuffle):
(JSC::Wasm::B3IRGenerator::addSIMDLoadSplat):
(JSC::Wasm::B3IRGenerator::addSIMDLoadLane):
(JSC::Wasm::B3IRGenerator::addSIMDStoreLane):
(JSC::Wasm::B3IRGenerator::addSIMDLoadExtend):
(JSC::Wasm::B3IRGenerator::addSIMDLoadPad):

Canonical link: <a href="https://commits.webkit.org/257865@main">https://commits.webkit.org/257865@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00d2f428ed4ab88594ff35fd34546b1fb900d10d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100163 "22 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33235 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109495 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10252 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92590 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107384 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105932 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7754 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91005 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34422 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77361 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90739 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3094 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23915 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86710 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/530 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3069 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29058 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9199 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43401 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89592 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5400 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4904 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20030 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/B3Const128Value.h:
```diff
@@ -0,0 +1,63 @@
+
+namespace JSC { namespace B3 {
+
+class JS_EXPORT_PRIVATE Const128Value : public Value {
```
Let's put `final`, then use `final` instead of `override`.

Constellation commented on Source/JavaScriptCore/b3/B3SIMDValue.cpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace JSC { namespace B3 {
+
+SIMDValue::~SIMDValue()
```
Use `= default;`.

Constellation commented on Source/JavaScriptCore/b3/B3Const128Value.cpp:
```diff
@@ -0,0 +1,44 @@
+
+Const128Value::~Const128Value()
+{
+}
```
Use `= default;`.

Constellation commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3702,6 +4054,18 @@ class LowerToAir {
             return;
         }
 
+        case Const128: {
+            auto value = m_value->as<Const128Value>()->value();
+            auto a = tmpForType(Int64);
+            auto result = tmp(m_value);
+            append(MoveZeroToVector, result);
+            append(Move, Arg::bigImm(value.u64x2[0]), a);
+            append(Air::VectorReplaceLaneInt64, Arg::imm(0), a, result);
+            append(Move, Arg::bigImm(value.u64x2[1]), a);
+            append(Air::VectorReplaceLaneInt64, Arg::imm(1), a, result);
+            return;
+        }
```
I think better way of this handling is making it a load from memory, as the same to `double` constant handling.
See B3MoveConstants.cpp. And then, here, we only use `MoveZeroToVector`.

justinmichaud replied:
Absolutely this would be better, this should be temporary

MenloDorian commented on Source/JavaScriptCore/b3/B3ArgumentRegValue.h:
```diff
@@ -33,13 +33,16 @@
 namespace JSC { namespace B3 {
 
 class JS_EXPORT_PRIVATE ArgumentRegValue final : public Value {
+    struct VectorTag { };
```
We usually use an enum for this like so:
```
enum VectorOpTag { VectorOp };
```
... and when calling `opcodeFromConstructor`, you would say `opcodeFromConstructor(origin, ArgumentRegValue::VectorOp)` instead.

MenloDorian commented on Source/JavaScriptCore/b3/B3ArgumentRegValue.h:
```diff
@@ -33,13 +33,16 @@
     Reg argumentReg() const { return m_reg; }
 
+    WTF_INTERNAL static constexpr VectorTag useFullRegister = { };
+
```
No need for this given the standard idiom (see above)?

MenloDorian commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -1227,6 +1225,87 @@ class LowerToAir {
         append(createStore(kind, memory->child(0), dest));
     }
 
+    Air::Opcode simdOpcode(SIMDLane lane, Air::Opcode i8, Air::Opcode i16, Air::Opcode i32, Air::Opcode i64, Air::Opcode f32, Air::Opcode f64)
```
nit: Since this is only called from one place, and the opcodes passed in are always constant, you can make this a template function:
```
template<Air::Opcode i8, Air::Opcode i16, Air::Opcode i32, Air::Opcode i64, Air::Opcode f32, Air::Opcode f64>
Air::Opcode simdOpcode(SIMDLane lane)
```
Might result in generating less code if the compiler isn't smart enough to begin with.

MenloDorian commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -1227,6 +1225,87 @@ class LowerToAir {
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    Air::Opcode simdOpcode(SIMDLane lane, SIMDSignMode signMode, Air::Opcode unsignedI8, Air::Opcode signedI8, Air::Opcode unsignedI16, Air::Opcode signedI16, Air::Opcode i32, Air::Opcode i64, Air::Opcode f32, Air::Opcode f64)
```
Ditto, can templatize this.

MenloDorian commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3619,6 +3698,279 @@ class LowerToAir {
+            Tmp shiftVector = m_code.newTmp(B3::FP);
+            append(And32, Arg::bitImm(mask), shift, shiftAmount);
+
+            if (isARM64()) {
```
Let's use `if constexpr (isARM64())`

MenloDorian commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3619,6 +3698,279 @@ class LowerToAir {
+                return;
+            }
+            
+            if (isX86()) {
```
Ditto.  Use `if constexpr (isX86())`

MenloDorian commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3619,6 +3698,279 @@ class LowerToAir {
+            emitSIMDCompare(Air::Arg::relCond(MacroAssembler::LessThanOrEqual), Air::Arg::doubleCond(MacroAssembler::DoubleLessThanOrEqualAndOrdered));
+            return;
+        case B3::VectorBelow:
+            emitSIMDCompare(Air::Arg::relCond(MacroAssembler::Below), Air::Arg());
```
Why not make a second `emitSIMDCompare` which only access a `relCond`?  Passing `Air::Arg()` feels like such a hack here.   Ditto below.

MenloDorian reviewed (approved):
r=me

Commit:
[SIMD] B3/OMG Support
https://bugs.webkit.org/show_bug.cgi?id=246347

Reviewed by Mark Lam.

Support B3/ the OMG tier for SIMD functions.

We add SimdValue and Const128Value as two new B3 value types, plus
opcodes for every wasm simd op.

We also add extra testing modes to simd spec tests to ensure that
the OMG tier is being fully exercised. This should be a temporary
measure until SIMD no longer requires tiering up from LLInt to BBQ.

Finally, we work around a bug in register allocation for SIMD functions
by disabling the use of ARM64 fp callee-saves.

* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::notifyFunctionUsesSIMD):
(JSC::Wasm::B3IRGenerator::addConstant):
(JSC::Wasm::B3IRGenerator::addExtractLane):
(JSC::Wasm::B3IRGenerator::addReplaceLane):
(JSC::Wasm::B3IRGenerator::addSIMDI_V):
(JSC::Wasm::B3IRGenerator::addSIMDV_V):
(JSC::Wasm::B3IRGenerator::addSIMDBitwiseSelect):
(JSC::Wasm::B3IRGenerator::addSIMDRelOp):
(JSC::Wasm::B3IRGenerator::addSIMDSwizzle):
(JSC::Wasm::B3IRGenerator::addSIMDV_VV):
(JSC::Wasm::B3IRGenerator::addSIMDLoad):
(JSC::Wasm::B3IRGenerator::addSIMDStore):
(JSC::Wasm::B3IRGenerator::addSIMDSplat):
(JSC::Wasm::B3IRGenerator::addSIMDShift):
(JSC::Wasm::B3IRGenerator::addSIMDExtmul):
(JSC::Wasm::B3IRGenerator::addSIMDShuffle):
(JSC::Wasm::B3IRGenerator::addSIMDLoadSplat):
(JSC::Wasm::B3IRGenerator::addSIMDLoadLane):
(JSC::Wasm::B3IRGenerator::addSIMDStoreLane):
(JSC::Wasm::B3IRGenerator::addSIMDLoadExtend):
(JSC::Wasm::B3IRGenerator::addSIMDLoadPad):

Canonical link: https://commits.webkit.org/257865@main
